### PR TITLE
[CPU] Introduce TilingConfig class

### PR DESCRIPTION
**Pull request by dcaballe**
TilingConfig is a simple steps towards separating the API to retrieve the tile size information from the actual representation and implementation of such information. It will let us implement different tiling configuration scenarios without exposing the implementation details or even replacing LoweringConfig with something else without impacting TilingConfig users.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: can we move this out of `switch`? It's duplicated many times below.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h:22]: Can we avoid using forward declaration, and include the header instead? Forward declaration can save compilation time for sure, but it hides the dependancy.

https://google.github.io/styleguide/cppguide.html#Forward_Declarations

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:445]: We don't want it be modified. Should this be `ArrayRef`?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: I think they should be initialized to `TilingLevel::InvalidLevel`? Maybe can just use `std::fill` here.

The invalid value check mismatch between here and l.175.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: I think it should just be `!=`. Using `<` assumes that invalid always has the largest value, which might introduce a bug accidentally.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: We prefer C++ casting (i.e., static_cast). It is fine to use `enum` to me because they are scoped to private members. Then we don't need casting at all.

**Review by hanhanW — CHANGES_REQUESTED** [review]

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: That was the first implementation but not all the pipelines require a lowering config so I can't create the TilingConfig outside of the switch. I can create a utility function with this two lines if that helps... Any other idea?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h:22]: I've never seen that rule being followed in MLIR/LLVM, quite strongly the opposite? I've seen fw declarations even with templates, etc. compile time is really a big deal. Are we following a different policy in IREE?

File: compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/Passes.cpp:445]: The information has to be materialized in memory as there is no original container with it. Perhaps in a follow-up iteration we may want to materialize some of the information in TilingConfig and then return references to it but currently TilingConfig is mostly a wrapper around LoweringConfig.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: Goot catch! I thought I was already doing that!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: Thanks! I made the enum unsigned. We can't use `!=` because that would make values > InvalidLevel to go through.

**Review by dcaballe — COMMENTED** [review]
Thanks for the quick review!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPULowerExecutableTarget.cpp]: I think we can allow empty TilingConfig as default. If the lowering_config is not provided, it basically returns empty lists for all the methods. Then it would be something like

```cpp
TilingConfig tilingConfig
if (succeeded(maybeLoweringConfig)) {
  tilingConfig.setConfig(*maybeLoweringConfig);
}
```

(IIUC, this can also address the failures in pipeline_tests.mlir.)

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: good point, SG!

File: compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h
Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h:22]: I think we don't have strong rules in MLIR/LLVM. It's common when there are cyclic deps. These files are complied together under `LLVMCPU` target, so I think it does not impact compilation time a lot -- if it is a pure header file. One question, why do we keep the implementation in the `.h` file but not `.cpp` file?

(I might have to take the statement back because the implementation is located in the header file.)

Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/LLVMCPUPasses.h:22]: Sorry, not a reason to keep things in .h. Moving the larger functions to the .cpp

File: compiler/src/iree/compiler/Codegen/LLVMCPU/BUILD.bazel
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/BUILD.bazel]: we should move `.h` to `hdrs`.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (hanhanW) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: should `tilingLevelToActualLevelMap` be initialized to `InvalidLevel` in this case?

**Review by hanhanW — APPROVED** [review]

**Review by MaheshRavishankar — COMMENTED** [review]
I am not sure what to make of this change. The `TilingConfig` is basically a wrapper around `loweringConfig`. So its moving from something that is already in IR to something that is not represented by a non-serialized/non-serializable data structure. So I am finding it hard to see why this helps? (Note: I am skeptical of this change, but I realize it might be a preference thing, so not blocking it as yet, but more asking questions).

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: Sorry, I left code uncommitted. The last commit didn't even build. I committed everything and addressed your comments, in case you want to take another quick look. Thanks!

**Comment by dcaballe** [comment]
This is a basic OOD principle. We provide a public interface and encapsulates the implementation behind the interface, hiding the complexity of the implementation from the user. This has quite a few advantages. For example... The user will always use the same interface while we can change the implementation and even the representation at IR level. This means, we could change, move around or evolve `LoweringConfig` and changes on the user side should be minimal. `TilingConfig` is also abstracting away the complexity of having multiple tiling scenarios, such as the ones with 3 tiling levels (for distribution and vectorization) and the ones for multi-level tiling with 5 levels (distribution, cache, vectorization) and providing verification for all of them in a single place, instead of scatter verification on the user's side. This will prevent us from, first, creating and accidentally using invalid scenarios that are not supported in LLVM CPU and might lead to unexpected behavior (e.g., one with 4 or 6 levels). This will prevent the user from misinterpreting the levels (i.e., no more `tilinglevels[numLevels - 2]`. If the user does `tilingConfig.getVectorParallelSizes()` be sure it will get the vector parallel sizes and not something else, regardless of the levels in the configuration. If the user tries to get a level that is not in the configuration it will get an error.

Not sure I understand the question about serialization. This is not intended to replace `LoweringConfig`.

Hopefully that helps.  This is part of the multi-tiling work that it's useful regardless.

**Comment by MaheshRavishankar** [comment]
Thanks. I am not blocking this change. Its already approved and will let you guys decide if this is the best way forward.

> This is a basic OOD principle. We provide a public interface and encapsulates the implementation behind the interface, hiding the complexity of the implementation from the user. This has quite a few advantages. For example... The user will always use the same interface while we can change the implementation and even the representation at IR level. This means, we could change, move around or evolve `LoweringConfig` and changes on the user side should be minimal. `TilingConfig` is also abstracting away the complexity of having multiple tiling scenarios, such as the ones with 3 tiling levels (for distribution and vectorization) and the ones for multi-level tiling with 5 levels (distribution, cache, vectorization) and providing verification for all of them in a single place, instead of scatter verification on the user's side. This will prevent us from, first, creating and accidentally using invalid scenarios that are not supported in LLVM CPU and might lead to unexpected behavior (e.g., one with 4 or 6 levels). This will prevent the user from misinterpreting the levels (i.e., no more `tilinglevels[numLevels - 2]`. If the user does `tilingConfig.getVectorParallelSizes()` be sure it will get the vector parallel sizes and not something else, regardless of the levels in the configuration. If the user tries to get a level that is not in the configuration it will get an error.

This makes sense.

> 
> Not sure I understand the question about serialization. This is not intended to replace `LoweringConfig`.

I think its a matter of debugability for me. Now there is a layer of indirection from what is in IR, to what is used in code. Maybe that indirection is useful, but I am just checking if the indirection is worth it. Again a preference maybe, but stating for record, and I am fine if this helps maintanence of the code overall.

> 
> Hopefully that helps. This is part of the multi-tiling work that it's useful regardless.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h
Comment (dcaballe) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/TileSizeSelection.h]: Unfortunately, this doesn't work as expected, because we create an invalid `TilingConfig` (without a valid `LowerginConfig`) and then we try to retrieve information from it using the invalid `LoweringConfig`. I think we shouldn't allow to create an invalid `TilingLowering`. I can go back to the previous state and move the repetitive code to a utility function.

**Comment by dcaballe** [comment]
> I think its a matter of debugability for me. Now there is a layer of indirection from what is in IR, to what is used in code. Maybe that indirection is useful, but I am just checking if the indirection is worth it. Again a preference maybe, but stating for record, and I am fine if this helps maintanence of the code overall.

That shouldn't be a problem as the actual `LoweringConfig` is there, e.g., `p tilingConfig.loweringConfig`

**Review by hanhanW — APPROVED** [review]

**Pull request merged**
